### ISSUE TITLE
Non-descriptive error on type mismatch for functions of scipy.optimize using zeros.h

### ISSUE THREAD

**Issue by Bocchio** [scipy/scipy#10236]
Many functions in scipy.optimize have an optional `maxiter` argument. This argument is expected to be an instance of int. Although that's in the documentation, the error when it fails isn't descriptive.

### Reproducing code example:
```
import scipy.optimize

scipy.optimize.bisect(lambda x: x, -1, 1, maxiter=100.0)
```

### Error message:

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3.7/site-packages/scipy/optimize/zeros.py", line 529, in bisect
    r = _zeros._bisect(f, a, b, xtol, rtol, maxiter, args, full_output, disp)
RuntimeError: Unable to parse arguments
```
Which says nothing about the problem and blames an internal function.
This is because
` scipy/scipy/optimize/Zeros/bisect.c ` expects an int for iter (and maxiter maps to iter)

```
double
bisect(callback_type f, double xa, double xb, double xtol, double rtol,
int iter, void *func_data, scipy_zeros_info *solver_stats)
```

It should be checked beforehand that `maxiter` is an instance of int to get a more userfriendly error message such as `TypeError: 10.0: expected int but got float`

**Comment by Bocchio** [comment]
I suggest adding:
```Python
if not isinstance(maxiter, int):
    raise TypeError("maxiter must be an integer, not %s" %
                     type(maxiter).__name__)
```
After 
```Python
if not isinstance(args, tuple):
    args = (args,)
```
For each function on scipy/optimize/zeros.py

**Comment by andyfaff** [comment]
Would it be easier to cast to an int?

maxiter = int (maxiter)

**Comment by ev-br** [comment]
`maxiter = operator.index(maxiter)`.

**Comment by Bocchio** [comment]
> Would it be easier to cast to an int? maxiter = int (maxiter)

I really think the function should raise an exception for cases in which `maxiter` is something like 10.5
This issue may arise silently because `int.__truediv__` returns a float. There isn't a type annotation either. So an user searching for roots (maybe using scipy indirectly, without knowing it) might adjust the iterations manually with code along the lines:
```
maxiter /= 10
```
And at no point will the traceback inform them that what failed was that argument.

**Comment by andyfaff** [comment]
But if the coercion is done in the Scipy function then passing a float
wouldn't result in an error.

**Comment by Bocchio** [comment]
I'm not arguing that, but in that case the documentation for maxiter should also change accordingly.
Current one just asks for an int.
```
maxiter : int, optional
        if convergence is not achieved in `maxiter` iterations, an error is
        raised.  Must be >= 0.
```
Personally I think that if a user pass something like `72.45`, maybe they don't know what they are doing and the best would be to inform them their mistake.
At least python's built-in class `range` took that path.

**Comment by Psirus** [comment]
This was fixed by #10907.

**Comment by tylerjereddy** [comment]
Yes, thanks, closing.
